rfc/rfc-87/rfc-87.md

```diff
@@ -24,9 +24,329 @@
+  }
+
+  @Override
+  public void swapKey(MemorySegment segI, int offsetI, MemorySegment segJ, int offsetJ) {
+    byte[] temp = new byte[NUM_KEY_BYTES];
+    segI.swapBytes(temp, segJ, offsetI, offsetJ, NUM_KEY_BYTES);
+  }
+
+  @Override
+  public int getNumKeyBytes() {
+    return NUM_KEY_BYTES;
+  }
+
+  @Override
+  public boolean isKeyFullyDetermines() {
+    return false;
+  }
+
+  @Override
+  public boolean invertKey() {
+    return false;
+  }
+}
+```
+
+Comparing record using _hoodie_record_key has a problem, for ComplexKeyGenerator _hoodie_record_key starts with column name.
+It makes BinaryInMemorySortBuffer usage inefficient, because BinaryInMemorySortBuffer builds index with first 8 bytes from key to make sorting faster (avoid records serialization/deserialization). 
+Here is possible to use something like FastRecordKeyComparator to trim column name from key values.
+
+```Java
+/**
+ * Comparator for record keys used in sorting operations before writing and during data compaction.
+ * Utilizes {@link SignedBytes#lexicographicalComparator()} for high-performance lexicographical comparison.
+ *
+ * <p>Writer uses this class for sorting data before writing, while the Compaction service uses it
+ * for determining the order of records.
+ */
+public class FastRecordKeyComparator implements Comparator<String> {
+  Comparator<byte[]> comparator = SignedBytes.lexicographicalComparator();
+
+  /**
+   * Compares two string representing full record keys.
+   *
+   * @return Negative if a < b, positive if a > b, 0 if equal.
+   */
+  @Override
+  public int compare(String recordKeyA, String recordKeyB) {
+    return compareBytesWithoutColumnName(
+        getRecordKeyBytes(recordKeyA),
+        getRecordKeyBytes(recordKeyB));
+  }
+
+  /**
+   * Compares two byte arrays representing record keys without column names and delimiters.
+   * 
+   * <p>These byte arrays should be a result of {@link #getRecordKeyBytes}
+   *
+   * @return Negative if a < b, positive if a > b, 0 if equal.
+   */
+  public int compareBytesWithoutColumnName(byte[] a, byte[] b) {
+    return comparator.compare(a, b);
+  }
+
+  /**
+   * Converts a record key into a byte array by extracting only values (without column name)
+   * and encoding them in UTF-8.
+   *
+   * @return Byte array representing record key without column names and delimiters, these byte arrays
+   * should only be compared using {@link #compareBytesWithoutColumnName(byte[], byte[])} method.  
+   */
+  public byte[] getRecordKeyBytes(String recordKey) {
+    return String.join("", KeyGenUtils.extractRecordKeys(recordKey))
+        .getBytes(StandardCharsets.UTF_8);
+  }
+}
+```
+
+### FlinkAppendHandle / FlinkCreateHandle
+
+The key idea here is to use Iterator provided by DataBucket instead of List<HoodieRecrod>.
+
+For FlinkAppendHandle, flushToDiskIfRequired looks redundant here, because storage and size calculation could be provided by DataBucket.
```

Comment:
> flushToDiskIfRequired looks redundant here

+1, the buffer management inside `FlinkAppendHandle` is unnecessary.  And actually, the current default buffer size for `flushToDiskIfRequired` is same with that of `DataBucket`, so for a log file for flink ingestion, there is always only one data block (with a deleta block if exists). 

I'm inclined to manage buffer size and flushing in one place, which will simplify writing logics, and if we want log blocks with small size, we can just decrease buffer size of `DataBucket`.

Reply:
The problem is the size estimation of data bucket could be not that accurate.

Reply:
Size estimation of data bucket correctly represents memory cost, I think, this is correct for write.batch.size.
Size estimation of result parquet log block could be performed only during writing process, so we can only limit log block size during writing.
Moreover, result log block size less than in-memory representation of records, for parquet with snappy it's about 5x times less than in-memory size, of course it depends on data.
In this question a agree with @cshuo

Reply:
I suggest to manage the buffer size in the DataBucket, it's so simple just calling getOccupancy() from BinaryInMemorySortBuffer.

Reply:
Okay